src/utils.c

```diff
@@ -1025,6 +1025,270 @@ ts_relation_size_impl(Oid relid)
 	return relsize;
 }
 
+/*
+ * Try to get cached size for a provided relation across all forks. The
+ * size is returned in terms of number of blocks.
+ *
+ * The function calls the underlying smgrnblocks if there is no cached
+ * data. That call populates the cache for subsequent invocations. This
+ * cached data gets removed asynchronously by PG relcache invalidations
+ * and then the refresh/cache cycle repeats till the next invalidation.
+ */
+static int64
+ts_try_relation_cached_size(Relation rel, bool verbose)
+{
+	BlockNumber result = 0, nblocks = 0;
+	ForkNumber forkNum;
+	bool cached = true;
+
+	/* Get heap size, including FSM and VM */
+	for (forkNum = 0; forkNum <= MAX_FORKNUM; forkNum++)
+	{
+#if PG14_LT
+		/* PG13 does not have smgr_cached_nblocks */
+		result = InvalidBlockNumber;
+#else
+		result = RelationGetSmgr(rel)->smgr_cached_nblocks[forkNum];
```

Comment:
How does this work if you pass in a foreign table or a view?

Reply:
Views and foreign tables do not have forks underneath. So both this function and `pg_relation_size` return 0 in those cases.

Reply:
Might be good to have a test for this so that there are no hidden bugs causing, for example, crashes.

Reply:
I see that you have such tests in the file below.